### ISSUE TITLE
[S017] Implement numerically stable row-wise softmax

### ISSUE THREAD

**Issue by 25d-228** [25d-228/gpt-2#19]
**Prerequisites:** S003 and S016; #1 applies. **Timebox:** 30–45 minutes.

## Learn and implement
Create `src/softmax.c` and `tests/test_softmax.c`; add its declaration to `include/ops.h`. For each finite, nonempty row, subtract its maximum, exponentiate, sum, and normalize. Operate independently across rows. Define an error for empty rows or nonfinite external inputs; masked attention will use only a finite valid prefix in S020. Explain why subtracting a constant preserves probabilities.

## Explicit tests
`[0,log(2),log(3)]` must give `[1/6,2/6,3/6]`; four equal logits give four probabilities of 0.25. `[1000,1001]` gives approximately `[0.26894142,0.73105858]` without overflow. Adding 100 to every logit must preserve results within tolerance. Test a singleton row, two distinct rows, sum approximately 1, nonnegative finite values, and rejection of NaN/Inf/empty input. Do not test probability normalization across the batch.
```sh
make test TEST=softmax
make sanitize TEST=softmax
make test
```
All pass with #1 tolerances. Temporarily remove max subtraction: the large-logit fixture must expose the failure.

## Done / teach-back
Explain logits versus probabilities and how normalization couples all entries in one row. No loss function, masking, or softmax backward in this lesson.
